clang/lib/Driver/Driver.cpp

```diff
@@ -3714,7 +3714,12 @@ class OffloadingActionBuilder final {
               auto *DeviceCheckAction =
                   C.MakeAction<SPIRCheckJobAction>(I, types::TY_Object);
               DeviceObjects.push_back(DeviceCheckAction);
-            } else {
+              continue;
+            }
+            // We want to move the AOCX/AOCR binary to the front of the objects
+            // allowing it to be picked up before the other device objects at
+            // runtime.
```

Comment:
Do I grasp the implication correctly here?
```suggestion
            // We want to move the AOCX/AOCR binary to the front of the objects
            // allowing it to be picked up instead of the other device objects at
            // runtime.
```
If so, it may be preferable to avoid confusion akin to "Wow, are we generating multiple device images now?" It's the path I've initially gone down, anyway :D

Reply:
I've also had that consideration of dropping the device compilation in the presence of the FPGA device binary.  Just having the FPGA device binary in the final image would solve this as well.  In the short term, however, I would like to go with this current solution with a TODO on improving behaviors when involving FPGA device.

Reply:
I see. Could it at least be improved so that the "extra" SPIR-V image is wrapped for `spir64`? Thus, we wouldn't have to rely on the implementation detail (limitation) of the image selection mechanism (picking the first available image), and a natural SPIR-V fallback would also become available. Wrapping a SPIR-V image as if it was an FPGA binary just feels too risky.

Reply:
That would definitely make things better in regards to image selection.  Things are very tightly tied to the device compilation triple and the handling of the AOCX binary is done somewhat separately so the amount of work involved to make the device compilation `spir64` wrapped is probably equivalent to the work needed to not perform the SPIR-V compile in the presense of AOCX.